File: app/src/main/java/com/example/paetz/yacguide/utils/FileChooser.java
package com.example.paetz.yacguide.utils;

import android.app.Dialog;
import android.content.Context;
import android.os.Environment;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.paetz.yacguide.R;

import java.io.File;
import java.io.FileFilter;
import java.util.Arrays;

public class FileChooser {

    private final String _PARENT_DIR = "..";

    private final Context _context;
    private ListView _list;
    private Dialog _dialog;
    private File _currentPath;

    public interface FileSelectedListener {
        void fileSelected(File file);
    }
    private FileSelectedListener _fileListener;

    public FileChooser setFileListener(FileSelectedListener fileListener) {
        _fileListener = fileListener;
        return this;
    }

    public FileChooser(Context context, String defaultFileName) {
        _context = context;
        _currentPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
        _dialog = new Dialog(context);
        _dialog.setContentView(R.layout.choose_file_dialog);

        final EditText fileNameEditText = (EditText) _dialog.findViewById(R.id.fileNameEditText);
        _list = (ListView) _dialog.findViewById(R.id.filesListView);
        _list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override public void onItemClick(AdapterView<?> parent, View view, int which, long id) {
                final File chosenFile = _getChosenFile((String) _list.getItemAtPosition(which));
                if (chosenFile.isDirectory()) {
                    _displayContent(chosenFile);
                } else {
                    fileNameEditText.setText(chosenFile.getName());
                }
            }
        });
        _dialog.findViewById(R.id.okButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String fileName = fileNameEditText.getText().toString();
                if (fileName.isEmpty()) {
                    Toast.makeText(_dialog.getContext(), "Keine Datei ausgewählt", Toast.LENGTH_SHORT).show();
                    return;
                } else if (!fileName.toLowerCase().endsWith(".json")) {
                    fileName += ".json";
                    Toast.makeText(_dialog.getContext(), "Dateisuffix \".json\" automatisch angefügt", Toast.LENGTH_SHORT).show();
                }
                if (_fileListener != null) {
                    _fileListener.fileSelected(new File(_currentPath, fileName));
                }
                _dialog.dismiss();
            }
        });
        _dialog.findViewById(R.id.cancelButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                _dialog.dismiss();
            }
        });
        _dialog.findViewById(R.id.chooseFileButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                _displayContent(_currentPath);
            }
        });

        ((TextView) _dialog.findViewById(R.id.currentPathTextView)).setText("Pfad: " + _currentPath.getAbsolutePath());
        fileNameEditText.setText(defaultFileName);
    }

    public void showDialog() {
        _dialog.show();
    }

    private void _displayContent(File path) {
        _currentPath = path;
        if (!_currentPath.exists()) {
            _currentPath.mkdirs();
        }
        File[] dirs = _currentPath.listFiles(new FileFilter() {
            @Override
            public boolean accept(File file) {
                return (file.canRead() && file.isDirectory());
            }
        });
        File[] files = _currentPath.listFiles(new FileFilter() {
            @Override public boolean accept(File file) {
                return (file.canRead() && !file.isDirectory() && file.getName().toLowerCase().endsWith(".json"));
            }
        });

        int i = 0;
        String[] fileList;
        if (_currentPath.getParentFile() == null || _currentPath.getParentFile().listFiles() == null) {
            fileList = new String[dirs.length + files.length];
        } else {
            fileList = new String[dirs.length + files.length + 1];
            fileList[i++] = _PARENT_DIR;
        }
        Arrays.sort(dirs);
        Arrays.sort(files);
        for (final File dir : dirs ) {
            fileList[i++] = dir.getName();
        }
        for (final File file : files ) {
            fileList[i++] = file.getName();
        }

        ((TextView) _dialog.findViewById(R.id.currentPathTextView)).setText("Pfad: " + _currentPath.getAbsolutePath());
        _list.setAdapter(new ArrayAdapter(_context,
                android.R.layout.simple_list_item_1, fileList) {
            @Override public View getView(int pos, View view, ViewGroup parent) {
                view = super.getView(pos, view, parent);
                ((TextView) view).setSingleLine(true);
                return view;
            }
        });
    }

    private File _getChosenFile(String fileName) {
        if (fileName.equals(_PARENT_DIR)) {
            return _currentPath.getParentFile();
        } else {
            return new File(_currentPath, fileName);
        }
    }
}

File: app/src/main/java/com/example/paetz/yacguide/utils/FileChooser.kt
package com.example.paetz.yacguide.utils

import android.app.Dialog
import android.content.Context
import android.os.Environment
import android.view.View
import android.view.ViewGroup
import android.widget.*

import com.example.paetz.yacguide.R

import java.io.File
import java.util.Arrays

class FileChooser(private val context: Context, defaultFileName: String) {

    private val parentDir = ".."
    private val list: ListView
    private val dialog: Dialog
    private var currentPath: File? = null
    private var fileListener: FileSelectedListener? = null

    interface FileSelectedListener {
        fun fileSelected(file: File)
    }

    fun setFileListener(fileListener: FileSelectedListener): FileChooser {
        this.fileListener = fileListener
        return this
    }

    init {
        currentPath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS)
        dialog = Dialog(context)
        dialog.setContentView(R.layout.choose_file_dialog)

        val fileNameEditText = dialog.findViewById<View>(R.id.fileNameEditText) as EditText
        list = dialog.findViewById<View>(R.id.filesListView) as ListView
        list.onItemClickListener = AdapterView.OnItemClickListener { _, _, which, _ ->
            val chosenFile = getChosenFile(list.getItemAtPosition(which) as String)
            if (chosenFile.isDirectory) {
                displayContent(chosenFile)
            } else {
                fileNameEditText.setText(chosenFile.name)
            }
        }
        dialog.findViewById<View>(R.id.okButton).setOnClickListener(View.OnClickListener {
            var fileName = fileNameEditText.text.toString()
            if (fileName.isEmpty()) {
                Toast.makeText(dialog.context, "Keine Datei ausgewählt", Toast.LENGTH_SHORT).show()
                return@OnClickListener
            } else if (!fileName.toLowerCase().endsWith(".json")) {
                fileName += ".json"
                Toast.makeText(dialog.context, "Dateisuffix \".json\" automatisch angefügt", Toast.LENGTH_SHORT).show()
            }
            if (fileListener != null) {
                fileListener!!.fileSelected(File(currentPath, fileName))
            }
            dialog.dismiss()
        })
        dialog.findViewById<View>(R.id.cancelButton).setOnClickListener { dialog.dismiss() }
        dialog.findViewById<View>(R.id.chooseFileButton).setOnClickListener { displayContent(currentPath) }

        dialog.findViewById<TextView>(R.id.currentPathTextView).text = "Pfad: " + currentPath!!.absolutePath
        fileNameEditText.setText(defaultFileName)
    }

    fun showDialog() {
        dialog.show()
    }

    private fun displayContent(path: File?) {
        currentPath = path
        if (!currentPath!!.exists()) {
            currentPath!!.mkdirs()
        }
        val dirs = currentPath!!.listFiles { file -> file.canRead() && file.isDirectory }
        val files = currentPath!!.listFiles { file -> file.canRead() && !file.isDirectory && file.name.toLowerCase().endsWith(".json") }

        var i = 0
        val fileList: Array<String>
        if (currentPath!!.parentFile == null || currentPath!!.parentFile.listFiles() == null) {
            fileList = Array(dirs.size + files.size) {""}
        } else {
            fileList = Array(dirs.size + files.size + 1) {""}
            fileList[i++] = parentDir
        }
        Arrays.sort(dirs)
        Arrays.sort(files)
        for (dir in dirs) {
            fileList[i++] = dir.name
        }
        for (file in files) {
            fileList[i++] = file.name
        }

        dialog.findViewById<TextView>(R.id.currentPathTextView).text = "Pfad: " + currentPath!!.absolutePath
        list.adapter = object : ArrayAdapter<String>(context,
                android.R.layout.simple_list_item_1, fileList) {
            override fun getView(pos: Int, view: View?, parent: ViewGroup): View {
                val result = super.getView(pos, view, parent) as TextView
                result.setSingleLine(true)
                return result
            }
        }
    }

    private fun getChosenFile(fileName: String): File {
        return if (fileName == parentDir) {
            currentPath!!.parentFile
        } else {
            File(currentPath, fileName)
        }
    }
}
